### PR TITLE
Move shortcode handling to reST compiler’s compile_string

### PR DESCRIPTION
**Pull request by Kwpolska**
This is an API change, and I’m well aware of that. I think we can allow it, because nobody else seems to be using it.

### CODE REVIEW

**Comment by felixfontein** [comment]
Fine for me. I've been using the `compile_string` function so far, but I don't mind it processing shortcodes as well.

File: nikola/plugin_categories.py [hunk 1]
```diff
@@ -303,6 +303,11 @@ def compile(self, source, dest, is_two_file=True, post=None, lang=None):
         # ignore `compile_html`
         self.compile_html(source, dest, is_two_file)
 
+    def compile_string(self, data, source_path=None, is_two_file=True, post=None, lang=None):
+        """Compile the source file into HTML strings (with shortcode support)."""
```
Comment (felixfontein) [inline nikola/plugin_categories.py]: Maybe the documentation should say that this has to return a tuple whose first component is the HTML fragment?

File: nikola/plugins/compile/html.py [hunk 2]
```diff
@@ -41,6 +41,10 @@ class CompileHtml(PageCompiler):
     name = "html"
     friendly_name = "HTML"
 
+    def compile_string(self, data, source_path=None, is_two_file=True, post=None, lang=None):
+        """Compile HTML into HTML strings, with shortcode support."""
+        return self.site.apply_shortcodes(data, with_dependencies=True, extra_context={'post': post})
```
Comment (felixfontein) [inline nikola/plugins/compile/html.py]: The `if not is_two_file:` part from `compile` should be moved here, too, shouldn't it?

**Comment by felixfontein** [comment]
About ipynb: looks ok, but neither I'm not familiar with its internal API nor do I use it in my blogs, so no idea how good it works...

File: nikola/plugins/compile/html.py (hunk 2 above)
Comment (Kwpolska) [inline nikola/plugins/compile/html.py]: Sure, fixed.

File: nikola/plugin_categories.py (hunk 1 above)
Comment (Kwpolska) [inline nikola/plugin_categories.py]: Done.

File: nikola/plugin_categories.py [hunk 3]
```diff
@@ -304,7 +304,10 @@ def compile(self, source, dest, is_two_file=True, post=None, lang=None):
         self.compile_html(source, dest, is_two_file)
 
     def compile_string(self, data, source_path=None, is_two_file=True, post=None, lang=None):
-        """Compile the source file into HTML strings (with shortcode support)."""
+        """Compile the source file into HTML strings (with shortcode support).
+
+        Returns a tuple of at least two elements (html string, shortcode dependencies).
```
Comment (felixfontein) [inline nikola/plugin_categories.py]: Someone might interpret this that shortcode dependencies should always be the second element in the list (which is not true). Should we change the implementations to always return this as second, or adjust the documentation?

Comment (Kwpolska) [inline nikola/plugin_categories.py]: > Returns a tuple of at least two elements: HTML string [0] and shortcode dependencies [last].

**Comment by felixfontein** [comment]
LGTM

**Comment by Kwpolska** [comment]
Merging, and let’s hope nobody else breaks. (It wasn’t a public API before anyway)

**Pull request merged**